src/report.rs

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Intel SGX Enclave report structures.
+
+use crate::parameters::{Attributes, Features, MiscSelect, Xfrm};
+
+use core::fmt::{self, Display};
+
+#[derive(Debug, PartialEq)]
+#[non_exhaustive]
+pub enum ReportError {
+    InvalidMiscSelect(u32),
+    InvalidFeatures(u64),
+    InvalidXfrm(u64),
+}
+
+impl Display for ReportError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ReportError::InvalidMiscSelect(value) => {
+                write!(f, "Invalid misc select: {}", value)
+            }
+            ReportError::InvalidFeatures(value) => {
+                write!(f, "Invalid features: {}", value)
+            }
+            ReportError::InvalidXfrm(value) => {
+                write!(f, "Invalid xfrm: {}", value)
+            }
+        }
+    }
+}
+
+/// The enclave report body.
+///
+/// For more information see the following documents:
+///
+/// [Intel® Software Guard Extensions (Intel® SGX) Data Center Attestation Primitives: ECDSA Quote Library API](https://download.01.org/intel-sgx/dcap-1.0/docs/SGX_ECDSA_QuoteGenReference_DCAP_API_Linux_1.0.pdf)
+///
+/// Table 5, A.4. Quote Format
+///
+/// [Intel® 64 and IA-32 Architectures Software Developer's Manual Volume 3 (3A, 3B, 3C & 3D): System Programming Guide](https://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-software-developer-vol-3d-part-4-manual.html)
+///
+/// Table 38-21. Layout of REPORT
+#[derive(Clone, Debug)]
+#[repr(C)]
+pub struct ReportBody {
+    cpusvn: [u8; 16],
+    miscselect: u32,
+    reserved1: [u8; 28],
+    // The attributes field is composed of the features and xfrm fields.
+    features: u64,
+    xfrm: u64,
+    mrenclave: [u8; 32],
+    reserved2: [u8; 32],
+    mrsigner: [u8; 32],
+    reserved3: [u8; 96],
+    isv_prodid: u16,
+    isv_svn: u16,
+    reserved4: [u8; 60],
+    reportdata: [u8; 64],
+}
+
+impl ReportBody {
+    /// The SVN (security version number) of the processor.
+    pub fn cpu_security_version(&self) -> [u8; 16] {
+        self.cpusvn
+    }
+
+    /// Bit vector specifying which extended features are saved to the MISC region of the
+    /// SSA frame when an AEX occurs.
+    pub fn misc_select(&self) -> Result<MiscSelect, ReportError> {
+        let misc_select = u32::from_le(self.miscselect);
```

Comment:
I'm curious, why is endianness a concern here? Aren't all Intel CPUs little-endian, and isn't SGX exclusively an Intel thing?

Reply:
We will be using this in steward which may run on non-intel machines and we would want to check these fields

Reply:
@npmccallum Should we remove these checks?

Reply:
@jarkkojs this is not `steward` crate, but rather a generic `sgx` crate and IMO potentially this logic could be useful on CPUs with different endianness

Reply:
also these are read-only functions of private fields so if steward is changing them it would probably be from somewhere else

Reply:
We cannot assume the endianness of the reader of this structure. The reader will receive this structure from a quote in little endian and will need to validate the signature without modifying any bytes. However, then the reader will have to semantically interpret the bytes. That last part is where the conversion from little endian to host endianness comes in.

Let's use byte arrays everywhere with functions to decode as host endianness. `Report` is manually aligned to 512, which is required. I think that's good enough.